controllers/spiaccesstokenbinding_controller.go

```diff
@@ -81,35 +85,47 @@ type SPIAccessTokenBindingReconciler struct {
 // SetupWithManager sets up the controller with the Manager.
 func (r *SPIAccessTokenBindingReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	r.syncer = sync.New(mgr.GetClient())
+	r.finalizers = finalizer.NewFinalizers()
+	if err := r.finalizers.Register(linkedSecretsFinalizerName, &linkedSecretsFinalizer{client: r.Client}); err != nil {
+		return fmt.Errorf("failed to register the linked secrets finalizer: %w", err)
+	}
+
 	err := ctrl.NewControllerManagedBy(mgr).
 		For(&api.SPIAccessTokenBinding{}).
-		Owns(&corev1.Secret{}).
 		Watches(&source.Kind{Type: &api.SPIAccessToken{}}, handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
 			kcpWorkspace := logicalcluster.From(o)
-
-			bindings := &api.SPIAccessTokenBindingList{}
-
 			ctx := context.TODO()
 			if !kcpWorkspace.Empty() {
 				ctx = logicalcluster.WithCluster(ctx, kcpWorkspace)
 			}
 
-			if err := r.Client.List(ctx, bindings, client.InNamespace(o.GetNamespace())); err != nil {
+			requests, err := r.filteredBindingsAsRequests(ctx, kcpWorkspace.String(), o.GetNamespace(), func(_ api.SPIAccessTokenBinding) bool { return true })
+			if err != nil {
 				spiAccessTokenBindingLog.Error(err, "failed to list SPIAccessTokenBindings while determining the ones linked to SPIAccessToken",
 					"SPIAccessTokenName", o.GetName(), "SPIAccessTokenNamespace", o.GetNamespace())
 				return []reconcile.Request{}
 			}
-			ret := make([]reconcile.Request, 0, len(bindings.Items))
-			for _, b := range bindings.Items {
-				ret = append(ret, reconcile.Request{
-					ClusterName: kcpWorkspace.String(),
-					NamespacedName: types.NamespacedName{
-						Name:      b.Name,
-						Namespace: b.Namespace,
-					},
-				})
+
+			return requests
+		})).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+			kcpWorkspace := logicalcluster.From(o)
+			ctx := context.TODO()
+			if !kcpWorkspace.Empty() {
+				ctx = logicalcluster.WithCluster(ctx, kcpWorkspace)
+			}
+
+			requests, err := r.filteredBindingsAsRequests(ctx, kcpWorkspace.String(), o.GetNamespace(), func(binding api.SPIAccessTokenBinding) bool {
+				// return binding.Status.SyncedObjectRef == toObjectRef(o)  //i wanted that. But version and kind of o is empty by some unknown reason
```

Comment:
This a handler is for Secrets... so the groupId is "", the version is "v1" and the kind is "Secret".

Reply:
I was logged that...no kind or version, only name is there (idfk why)

Reply:
yes, but you're watching the secrets, so that handler is receiving the secrets that you asked for, right? So even though `o` doesn't have the kind and version, you can be sure it is the secrets, because that's what that handler function is configured to handle.